src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_MAX_SIZE = 10;
+    private static final int MIN_SIZE = 0;
+    private T[] data;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_MAX_SIZE];
+        size = MIN_SIZE;
+    }
+
     @Override
     public void add(T value) {
-
+        growIfFull();
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkAddRange(index);
+        growIfFull();
+        System.arraycopy(data, index, data, index + 1, size - index);
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int addedArraySize = list.size();
+        while (addedArraySize > data.length - size) {
+            grow();
+        }
+        for (int i = 0; i < addedArraySize; i++) {
+            data[size + i] = list.get(i);
+        }
+        size += addedArraySize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkRange(index);
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkRange(index);
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkRange(index);
+        T value = data[index];
+        if (size - 1 > index) {
+            System.arraycopy(data, index + 1, data, index, size - 1);
+        }
+        size--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (equals(data[i], element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException(element.toString() + ": doesn't exist");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public boolean equals(T first, T second) {
+        return (first == second) || (first != null && first.equals(second));
+    }
+
+    private void grow() {
+        int oldSize = data.length;
+        int newSize = oldSize + (oldSize >> 1);
```

Comment:
please, do not use the magic numbers 1 => create new private static final double GROWING_COEFFICIENT = 1.5d and use it here